stdlib/public/runtime/GenericMetadataBuilder.cpp

```diff
@@ -0,0 +1,408 @@
+  }
+}
+
+template <typename T>
+static const T &unwrapVWTField(const T &field) {
+  return field;
+}
+
+static uint32_t unwrapVWTField(const ValueWitnessFlags &field) {
+  return field.getOpaqueValue();
+}
+
+static bool equalVWTs(const ValueWitnessTable *a, const ValueWitnessTable *b) {
+#define WANT_ONLY_REQUIRED_VALUE_WITNESSES
+#define FUNCTION_VALUE_WITNESS(LOWER_ID, UPPER_ID, RET, PARAMS)                \
+  if (a->LOWER_ID != b->LOWER_ID)                                              \
+    return false;
+#define VALUE_WITNESS(LOWER_ID, UPPER_ID)                                      \
+  if (unwrapVWTField(a->LOWER_ID) != unwrapVWTField(b->LOWER_ID))              \
+    return false;
+#include "swift/ABI/ValueWitness.def"
+
+  auto *enumA = dyn_cast<EnumValueWitnessTable>(a);
+  auto *enumB = dyn_cast<EnumValueWitnessTable>(b);
+  if (enumA == nullptr && enumB == nullptr) {
+    return true;
+  } else if (enumA != nullptr && enumB != nullptr) {
+#define WANT_ONLY_ENUM_VALUE_WITNESSES
+#define VALUE_WITNESS(LOWER_ID, UPPER_ID)                                      \
+  if (unwrapVWTField(enumA->LOWER_ID) != unwrapVWTField(enumB->LOWER_ID))      \
+    return false;
+#include "swift/ABI/ValueWitness.def"
+    return true;
+  } else {
+    // Only one of a and b is an enum table.
+    return false;
+  }
+}
+
+void swift::validateExternalGenericMetadataBuilder(
+    const Metadata *original, const TypeContextDescriptor *description,
+    const void *arguments) {
+  if (auto valueDescriptor = dyn_cast<ValueTypeDescriptor>(description)) {
+    if (valueDescriptor->isGeneric()) {
+      auto pattern = reinterpret_cast<GenericValueMetadataPattern *>(
+          valueDescriptor->getFullGenericContextHeader()
+              .DefaultInstantiationPattern.get());
+      auto extraDataSize = genericValueDataExtraSize(valueDescriptor, pattern);
+      if (auto *error = extraDataSize.getError()) {
+        printf("Failed to get extra data size: %s\n", error->cStr());
+        return;
+      }
+
+      auto newMetadata = allocateGenericValueMetadata(
+          valueDescriptor, arguments, pattern, *extraDataSize.getValue());
+      if (!newMetadata)
+        return;
+      bool success = initializeGenericValueMetadata(newMetadata);
+      if (!success)
+        return;
+
+      auto origVWT = asFullMetadata(original)->ValueWitnesses;
+      auto newVWT = asFullMetadata(newMetadata)->ValueWitnesses;
+
+      bool equal = true;
+      if (!equalVWTs(origVWT, newVWT)) {
+        printf("VWTs do not match\n");
+        equal = false;
+      }
+      size_t totalSize = sizeof(ValueMetadata) + *extraDataSize.getValue();
+      if (memcmp(original, newMetadata, totalSize)) {
+        printf("Metadatas do not match\n");
+        equal = false;
+      }
+
+      if (!equal) {
+        printf("Error! Mismatch between new/old metadata builders!\n");
+        printf("Original metadata:\n");
+        _swift_dumpMetadata(original);
```

Comment:
(If we change these from `printf`s to anything else, we probably also need to update `_swift_dumpMetadata()` in a similar way.)